# Celebal Technologies - Week 3 SQL Assignment
## Advanced SQL Analysis Using Subqueries, CTEs, and Window Functions
### Submitted By:
Amit Singh

# Step 1: Import Required Libraries


In [3]:
import pandas as pd
import sqlite3

## Step 2: Load Superstore Dataset


In [4]:
df = pd.read_csv("data/Sample - Superstore.csv", encoding='latin1')

In [5]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,...,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,...,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2016-138688,6/12/2016,6/16/2016,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,...,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,...,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [6]:
df.shape

(9994, 21)

In [7]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country', 'City', 'State',
       'Postal Code', 'Region', 'Product ID', 'Category', 'Sub-Category',
       'Product Name', 'Sales', 'Quantity', 'Discount', 'Profit'],
      dtype='str')

In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   str    
 2   Order Date     9994 non-null   str    
 3   Ship Date      9994 non-null   str    
 4   Ship Mode      9994 non-null   str    
 5   Customer ID    9994 non-null   str    
 6   Customer Name  9994 non-null   str    
 7   Segment        9994 non-null   str    
 8   Country        9994 non-null   str    
 9   City           9994 non-null   str    
 10  State          9994 non-null   str    
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   str    
 13  Product ID     9994 non-null   str    
 14  Category       9994 non-null   str    
 15  Sub-Category   9994 non-null   str    
 16  Product Name   9994 non-null   str    
 17  Sales          9994 non-null   float64
 18  Quantity       9994

### Insight

The dataset includes important business information such as customer details, product information, sales records, profit data, discounts, and regional performance

# Step 3: Create SQL Database

A

In [9]:
conn = sqlite3.connect("superstore.db")

## Step 4: Load Raw Dataset into SQL Table

The complete dataset is stored inside a SQL table named `superstore_raw`.

In [10]:
df.to_sql(
    "superstore_raw",
    conn,
    if_exists="replace",
    index=False
)

9994

In [19]:
query = """
SELECT *
FROM superstore_raw
LIMIT 5
"""
conn.execute(query)

# Step 5: Create Customers,Product,Orders Table



In [17]:
query = """
CREATE TABLE customers AS
SELECT DISTINCT
    "Customer ID" AS customer_id,
    "Customer Name" AS customer_name
FROM superstore_raw
"""
conn.execute(query)

In [18]:
pd.read_sql("SELECT * FROM customers LIMIT 5", conn)

,customer_id,customer_name
0,CG-12520,Claire Gute
1,DV-13045,Darrin Van Huff
2,SO-20335,Sean O'Donnell
3,BH-11710,Brosina Hoffman
4,AA-10480,Andrew Allen


### Insight

The customers table contains unique customer records and removes redundancy from the original dataset.

In [ ]:
query = """
CREATE TABLE products AS
SELECT DISTINCT
    "Product ID" AS product_id,
    "Product Name" AS product_name,
    Category,
"Sub-Category" AS sub_category
FROM superstore_raw
"""
conn.execute(query)

In [16]:
pd.read_sql("SELECT * FROM products LIMIT 5", conn)

,product_id,product_name,Category,sub_category
0,FUR-BO-10001798,Bush Somerset Collection Bookcase,Furniture,Bookcases
1,FUR-CH-10000454,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",Furniture,Chairs
2,OFF-LA-10000240,Self-Adhesive Address Labels for Typewriters b...,Office Supplies,Labels
3,FUR-TA-10000577,Bretford CR4500 Series Slim Rectangular Table,Furniture,Tables
4,OFF-ST-10000760,Eldon Fold 'N Roll Cart System,Office Supplies,Storage


In [17]:
query = """
CREATE TABLE orders AS
SELECT DISTINCT
    "Order ID" AS order_id,
    "Order Date" AS order_date,
    "Ship Date" AS ship_date,
    Sales,
    Quantity,
    Discount,
    Profit,
    "Customer ID" AS customer_id,
    "Product ID" AS product_id
FROM superstore_raw
"""
conn.execute(query)

In [19]:
pd.read_sql("SELECT * FROM orders LIMIT 5",conn)

,order_id,order_date,ship_date,Sales,Quantity,Discount,Profit,customer_id,product_id
0,CA-2016-152156,11/8/2016,11/11/2016,261.9600,2,0.00,41.9136,CG-12520,FUR-BO-10001798
1,CA-2016-152156,11/8/2016,11/11/2016,731.9400,3,0.00,219.5820,CG-12520,FUR-CH-10000454
2,CA-2016-138688,6/12/2016,6/16/2016,14.6200,2,0.00,6.8714,DV-13045,OFF-LA-10000240
3,US-2015-108966,10/11/2015,10/18/2015,957.5775,5,0.45,-383.0310,SO-20335,FUR-TA-10000577
4,US-2015-108966,10/11/2015,10/18/2015,22.3680,2,0.20,2.5164,SO-20335,OFF-ST-10000760


# Step 6: Subqueries


Sales greater than the average sales


In [20]:
query = """
SELECT * FROM orders
WHERE Sales >
(SELECT AVG(Sales)
 FROM orders
)
"""
pd.read_sql(query, conn)

,order_id,order_date,ship_date,Sales,Quantity,Discount,Profit,customer_id,product_id
0,CA-2016-152156,11/8/2016,11/11/2016,261.9600,2,0.00,41.9136,CG-12520,FUR-BO-10001798
1,CA-2016-152156,11/8/2016,11/11/2016,731.9400,3,0.00,219.5820,CG-12520,FUR-CH-10000454
2,US-2015-108966,10/11/2015,10/18/2015,957.5775,5,0.45,-383.0310,SO-20335,FUR-TA-10000577
3,CA-2014-115812,6/9/2014,6/14/2014,907.1520,6,0.20,90.7152,BH-11710,TEC-PH-10002275
4,CA-2014-115812,6/9/2014,6/14/2014,1706.1840,9,0.20,85.3092,BH-11710,FUR-TA-10001539
...,...,...,...,...,...,...,...,...,...
2354,US-2016-103674,12/6/2016,12/10/2016,271.9600,5,0.20,27.1960,AP-10720,TEC-PH-10004080
2355,US-2016-103674,12/6/2016,12/10/2016,249.5840,2,0.20,31.1980,AP-10720,TEC-PH-10002496
2356,US-2016-103674,12/6/2016,12/10/2016,437.4720,14,0.20,153.1152,AP-10720,OFF-BI-10002026
2357,CA-2017-121258,2/26/2017,3/3/2017,258.5760,2,0.20,19.3932,DB-13060,TEC-PH-10003645


This query identifies orders whose sales value is greater than the overall average sales amount.

HIGHEST ORDER PER CUSTOMER

In [49]:
query = """
SELECT * FROM orders t1
WHERE Sales =
(SELECT MAX(Sales)
FROM orders
WHERE customer_id = t1.customer_id
)
"""
pd.read_sql(query, conn)

,order_id,order_date,ship_date,Sales,Quantity,Discount,Profit,customer_id,product_id
0,CA-2016-152156,11/8/2016,11/11/2016,731.9400,3,0.00,219.5820,CG-12520,FUR-CH-10000454
1,US-2015-108966,10/11/2015,10/18/2015,957.5775,5,0.45,-383.0310,SO-20335,FUR-TA-10000577
2,CA-2014-115812,6/9/2014,6/14/2014,1706.1840,9,0.20,85.3092,BH-11710,FUR-TA-10001539
3,CA-2015-106320,9/25/2015,9/30/2015,1044.6300,3,0.00,240.2649,EB-13870,FUR-TA-10000577
4,US-2015-150630,9/17/2015,9/21/2015,3083.4300,7,0.50,-1665.0522,TB-21520,FUR-BO-10004834
...,...,...,...,...,...,...,...,...,...
790,CA-2015-159534,3/20/2015,3/23/2015,1087.9360,8,0.20,353.5792,DH-13075,OFF-BI-10003656
791,CA-2016-129630,9/4/2016,9/4/2016,2799.9600,5,0.20,944.9865,IM-15055,TEC-CO-10003763
792,CA-2017-121559,6/1/2017,6/3/2017,2405.2000,8,0.00,793.7160,HW-14935,OFF-AP-10002945
793,CA-2017-153871,12/11/2017,12/17/2017,735.9800,2,0.00,331.1910,RB-19435,OFF-BI-10004600


# Step 7: Common Table Expressions (CTEs)


TOTAL SALES PER CUSTOMER

In [51]:
query = """
WITH customer_sales AS
(SELECT
    customer_id,
    ROUND(SUM(Sales),2) AS total_sales
    FROM orders
    GROUP BY customer_id
)
SELECT *
FROM customer_sales
ORDER BY total_sales DESC
"""
pd.read_sql(query, conn)

,customer_id,total_sales
0,SM-20320,25043.05
1,TC-20980,19052.22
2,RB-19360,15117.34
3,TA-21385,14595.62
4,AB-10105,14473.57
...,...,...
788,RS-19870,22.33
789,MG-18205,16.74
790,CJ-11875,16.52
791,LD-16855,5.30


The CTE calculates total sales generated by each customer.

# Step 8: Window Functions



In [27]:
query = """
SELECT customer_id, Sales,
ROW_NUMBER() OVER(ORDER BY Sales DESC)
AS row_num
FROM orders
"""
pd.read_sql(query, conn) 

,customer_id,Sales,row_num
0,SM-20320,22638.480,1
1,TC-20980,17499.950,2
2,RB-19360,13999.960,3
3,TA-21385,11199.968,4
4,HL-15040,10499.970,5
...,...,...,...
9988,HG-14965,0.876,9989
9989,JO-15280,0.852,9990
9990,KB-16600,0.836,9991
9991,RS-19765,0.556,9992


In [28]:
query = """
SELECT customer_id, Sales,
RANK() OVER( ORDER BY Sales DESC)
AS sales_rank
FROM orders
"""
pd.read_sql(query, conn)

,customer_id,Sales,sales_rank
0,SM-20320,22638.480,1
1,TC-20980,17499.950,2
2,RB-19360,13999.960,3
3,TA-21385,11199.968,4
4,HL-15040,10499.970,5
...,...,...,...
9988,HG-14965,0.876,9989
9989,JO-15280,0.852,9990
9990,KB-16600,0.836,9991
9991,RS-19765,0.556,9992


In [29]:
query = """
SELECT customer_id, Sales,
DENSE_RANK() OVER( ORDER BY Sales DESC)
AS sales_rank
FROM orders
"""
pd.read_sql(query, conn)

,customer_id,Sales,sales_rank
0,SM-20320,22638.480,1
1,TC-20980,17499.950,2
2,RB-19360,13999.960,3
3,TA-21385,11199.968,4
4,HL-15040,10499.970,5
...,...,...,...
9988,HG-14965,0.876,5821
9989,JO-15280,0.852,5822
9990,KB-16600,0.836,5823
9991,RS-19765,0.556,5824


top 3 customers based on total sales

In [ ]:
query = """
WITH customer_sales AS
(
SELECT customer_id,ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT *
FROM
(
SELECT customer_id, total_sales,
RANK() OVER(ORDER BY total_sales DESC )
AS sales_rank
FROM customer_sales
)
WHERE sales_rank <= 3;"""
pd.read_sql(query,conn)

,customer_id,total_sales,sales_rank
0,SM-20320,25043.05,1
1,TC-20980,19052.22,2
2,RB-19360,15117.34,3


# Step 9: Combined Business Analysis
JOIN + CTE + WINDOW FUNCTION

In [52]:
query = """
WITH customer_sales AS
( SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
c.customer_name,
cs.total_sales,
RANK() OVER(ORDER BY cs.total_sales DESC)
AS customer_rank
FROM customer_sales cs
JOIN customers c
ON cs.customer_id = c.customer_id
"""
pd.read_sql(query, conn)

,customer_name,total_sales,customer_rank
0,Sean Miller,25043.05,1
1,Tamara Chand,19052.22,2
2,Raymond Buch,15117.34,3
3,Tom Ashbrook,14595.62,4
4,Adrian Barton,14473.57,5
...,...,...,...
788,Roy Skaria,22.33,789
789,Mitch Gastineau,16.74,790
790,Carl Jackson,16.52,791
791,Lela Donovan,5.30,792



This query ranks customers based on total sales contribution 

top customers

In [53]:
query = """
WITH customer_sales AS
(
SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
t1.customer_name,
t2.total_sales
FROM customer_sales t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
ORDER BY t2.total_sales DESC
LIMIT 10
"""
pd.read_sql(query, conn)

,customer_name,total_sales
0,Sean Miller,25043.05
1,Tamara Chand,19052.22
2,Raymond Buch,15117.34
3,Tom Ashbrook,14595.62
4,Adrian Barton,14473.57
5,Ken Lonsdale,14175.23
6,Sanjit Chand,14142.33
7,Hunter Lopez,12873.30
8,Sanjit Engle,12209.44
9,Christopher Conant,12129.07


lowest customers

In [54]:
query = """
WITH customer_sales AS
(
SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
t1.customer_name,
t2.total_sales
FROM customer_sales t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
ORDER BY t2.total_sales ASC
LIMIT 10
"""
pd.read_sql(query, conn)

,customer_name,total_sales
0,Thais Sissman,4.83
1,Lela Donovan,5.30
2,Carl Jackson,16.52
3,Mitch Gastineau,16.74
4,Roy Skaria,22.33
5,Susan Gilcrest,47.95
6,Ricardo Emerson,48.36
7,Larry Blacks,50.19
8,Adrian Shami,58.82
9,Jasper Cacioppo,71.26


SINGLE ORDER CUSTOMERS

In [55]:
query = """
SELECT
t2.customer_name,
COUNT(t1.order_id) AS total_orders
FROM orders t1
JOIN customers t2
ON t1.customer_id = t2.customer_id
GROUP BY t1.customer_id
HAVING COUNT(t1.order_id) = 1
"""

pd.read_sql(query, conn)

,customer_name,total_orders
0,Anthony O'Donnell,1
1,Carl Jackson,1
2,Jocasta Rupert,1
3,Lela Donovan,1
4,Ricardo Emerson,1


# Step 10: Mini Project - Customer Sales Insights

This section answers real-world business questions using advanced SQL concepts such as CTEs, Subqueries, JOINs, and Window Functions.

### Query 1: Top 5 Customers Based on Total Sales

In [57]:
query = """
WITH customer_sales AS
(SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
t1.customer_name,
t2.total_sales
FROM customer_sales t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
ORDER BY t2.total_sales DESC
LIMIT 5
"""

pd.read_sql(query, conn)

,customer_name,total_sales
0,Sean Miller,25043.05
1,Tamara Chand,19052.22
2,Raymond Buch,15117.34
3,Tom Ashbrook,14595.62
4,Adrian Barton,14473.57


### Query 2: Bottom 5 Customers Based on Total Sales

In [58]:
query = """
WITH customer_sales AS
(
SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
t1.customer_name,
t2.total_sales
FROM customer_sales t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
ORDER BY t2.total_sales ASC

LIMIT 5
"""

pd.read_sql(query, conn)

,customer_name,total_sales
0,Thais Sissman,4.83
1,Lela Donovan,5.30
2,Carl Jackson,16.52
3,Mitch Gastineau,16.74
4,Roy Skaria,22.33


### Query 3: Customers Who Made Only One Order

In [59]:
query = """
SELECT
t1.customer_name,
COUNT(t2.order_id) AS total_orders
FROM orders t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
GROUP BY t2.customer_id
HAVING COUNT(t2.order_id) = 1
"""

pd.read_sql(query, conn)

,customer_name,total_orders
0,Anthony O'Donnell,1
1,Carl Jackson,1
2,Jocasta Rupert,1
3,Lela Donovan,1
4,Ricardo Emerson,1


### Query 4: Customers With Above-Average Sales

In [60]:
query = """
WITH customer_sales AS
(
SELECT customer_id, ROUND(SUM(Sales),2) AS total_sales
FROM orders
GROUP BY customer_id
)
SELECT
t1.customer_name,
t2.total_sales
FROM customer_sales t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
WHERE t2.total_sales >
(
SELECT AVG(total_sales)
FROM customer_sales
)
ORDER BY t2.total_sales DESC
"""

pd.read_sql(query, conn)

,customer_name,total_sales
0,Sean Miller,25043.05
1,Tamara Chand,19052.22
2,Raymond Buch,15117.34
3,Tom Ashbrook,14595.62
4,Adrian Barton,14473.57
...,...,...
289,Julie Kriz,2932.48
290,Shaun Weien,2921.54
291,Maris LaWare,2921.50
292,Rob Dowd,2912.89


### Query 5: Highest Order Value Per Customer

In [61]:
query = """
SELECT
t1.customer_name,
MAX(t2.Sales) AS highest_order_value
FROM orders t2
JOIN customers t1
ON t2.customer_id = t1.customer_id
GROUP BY t2.customer_id
ORDER BY highest_order_value DESC
"""
pd.read_sql(query, conn)

,customer_name,highest_order_value
0,Sean Miller,22638.480
1,Tamara Chand,17499.950
2,Raymond Buch,13999.960
3,Tom Ashbrook,11199.968
4,Hunter Lopez,10499.970
...,...,...
788,Carl Jackson,16.520
789,Mitch Gastineau,12.320
790,Roy Skaria,9.648
791,Lela Donovan,5.304


# Key Business Insights

- Sean Miller was identified as the highest revenue-generating customer with total sales of approximately $25,043, followed by Tamara Chand and Raymond Buch.
- Around 2,360 out of 9,994 order records had sales greater than the average sales value, indicating that a smaller number of high-value transactions contribute significantly to overall revenue.
- Only 294 out of 793 customers generated above-average total sales, showing that business revenue is concentrated among a limited group of customers.
- The analysis identified 12 single-order customers, highlighting potential opportunities for customer retention and re-engagement strategies.
- Customers such as Thais Sissman contributed very low sales values (below $5), demonstrating the presence of a long-tail customer distribution.
- Advanced SQL concepts including Subqueries, CTEs, and Window Functions helped perform efficient customer ranking, aggregation, and business analysis on the dataset.